library/std/src/env.rs

```diff
@@ -872,19 +874,28 @@ impl !Sync for Args {}
 #[stable(feature = "env", since = "1.0.0")]
 impl Iterator for Args {
     type Item = String;
+
     fn next(&mut self) -> Option<String> {
         self.inner.next().map(|s| s.into_string().unwrap())
     }
+
+    #[inline]
     fn size_hint(&self) -> (usize, Option<usize>) {
         self.inner.size_hint()
     }
+
+    // Methods which skip args cannot simply delegate to the inner iterator,
+    // because UTF-8 validation needs to be performed for every argument.
```

Comment:
This contract seems more ambiguous to me so I don't know that we should state it is clearly decided one way or another, even internally. I agree with not skipping for now, though, as deciding one way or another would still obviously cause externally observable changes, which would be a notable API update, which means it's a different PR even if we want to do that.

```suggestion
    // because `env::args` states that we will panic if an argument is not valid UTF-8
    // "during iteration". This offers two possible interpretations:
    // - a skipped argument is never encountered "during iteration"
    // - even a skipped argument is encountered "during iteration"
    // As a panic can be observed, we err towards checking even skipped arguments.
```

Reply:
Good wording. I've adopted it, slightly adjusted.